src/System.Linq/src/System/Linq/AppendPrepend.cs

```diff
@@ -0,0 +1,472 @@
+                }
+            }                
+
+            public override AppendPrependIterator<TSource> Prepend(TSource element)
+            {
+                if (_appending)
+                {
+                    return new AppendPrependNIterator<TSource>(_source, new SharedCollection<TSource>(element), 1, new SharedCollection<TSource>(_element), 1);
+                }
+                else
+                {
+                    return new AppendPrependNIterator<TSource>(_source, new SharedCollection<TSource>(element, _element), 2, null, 0);
+                }
+            }
+
+            public override int GetCount(bool onlyIfCheap)
+            {
+                IIListProvider<TSource> listProv = _source as IIListProvider<TSource>;
+                if (listProv != null)
+                {
+                    int count = listProv.GetCount(onlyIfCheap);
+                    return count == -1 ? -1 : count + 1;
+                }
+
+                return !onlyIfCheap || _source is ICollection<TSource> || _source is ICollection ? _source.Count() + 1 : -1;
+            }
+
+            public override TSource[] ToArray()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                if (count == -1)
+                {
+                    return EnumerableHelpers.ToArray(this);
+                }
+
+                TSource[] array = new TSource[count];
+                int index;
+                if (_appending)
+                {
+                    index = 0;
+                }
+                else
+                {
+                    array[0] = _element;
+                    index = 1;
+                }
+
+                foreach (TSource item in _source)
+                {
+                    array[index] = item;
+                    ++index;
+                }
+
+                if (_appending)
+                {
+                    Debug.Assert(index == array.Length - 1);
+                    array[index] = _element;
+                }
+
+                return array;
+            }
+
+            public override List<TSource> ToList()
+            {
+                int count = GetCount(onlyIfCheap: true);
+                List<TSource> list = count == -1 ? new List<TSource>() : new List<TSource>(count);
+                if (!_appending)
+                {
+                    list.Add(_element);
+                }
+
+                list.AddRange(_source);
+
+                if (_appending)
+                {
+                    list.Add(_element);
+                }
+
+                return list;
+            }
+        }
+
+        private sealed class AppendPrependNIterator<TSource> : AppendPrependIterator<TSource>
+        {
+            private readonly SharedCollection<TSource> _prepended;
+            private readonly SharedCollection<TSource> _appended;
```

Comment:
It's not so much that they're always copied when accessing them that would be the big problem, as they won't be copied back because they're readonly which would make it plain not work. Again though, we need a reference type here because we want to deliberately alias the same instance between different iterator instances.